src/compiler/tsbuild.ts

```diff
@@ -0,0 +1,1222 @@
+            if (f in lookup) {
+                return lookup[f];
+            }
+            else {
+                throw new Error(`No value corresponding to ${fileName} exists in this map`);
+            }
+        }
+
+        function getValueOrUndefined(fileName: string): T | undefined {
+            const f = normalizePath(fileName);
+            if (f in lookup) {
+                return lookup[f];
+            }
+            else {
+                return undefined;
+            }
+        }
+    }
+
+    export function createDependencyMapper() {
+        const childToParents: { [key: string]: ResolvedConfigFileName[] } = {};
+        const parentToChildren: { [key: string]: ResolvedConfigFileName[] } = {};
+        const allKeys: ResolvedConfigFileName[] = [];
+
+        function addReference(childConfigFileName: ResolvedConfigFileName, parentConfigFileName: ResolvedConfigFileName): void {
+            addEntry(childToParents, childConfigFileName, parentConfigFileName);
+            addEntry(parentToChildren, parentConfigFileName, childConfigFileName);
+        }
+
+        function getReferencesTo(parentConfigFileName: ResolvedConfigFileName): ResolvedConfigFileName[] {
+            return parentToChildren[normalizePath(parentConfigFileName)] || [];
+        }
+
+        function getReferencesOf(childConfigFileName: ResolvedConfigFileName): ResolvedConfigFileName[] {
+            return childToParents[normalizePath(childConfigFileName)] || [];
+        }
+
+        function getKeys(): ReadonlyArray<ResolvedConfigFileName> {
+            return allKeys;
+        }
+
+        function addEntry(mapToAddTo: typeof childToParents | typeof parentToChildren, key: ResolvedConfigFileName, element: ResolvedConfigFileName) {
+            key = normalizePath(key) as ResolvedConfigFileName;
+            element = normalizePath(element) as ResolvedConfigFileName;
+            const arr = (mapToAddTo[key] = mapToAddTo[key] || []);
+            if (arr.indexOf(element) < 0) {
+                arr.push(element);
+            }
+            if (allKeys.indexOf(key) < 0) allKeys.push(key);
+            if (allKeys.indexOf(element) < 0) allKeys.push(element);
+        }
+
+        return {
+            addReference,
+            getReferencesTo,
+            getReferencesOf,
+            getKeys
+        };
+    }
+
+    function getOutputDeclarationFileName(inputFileName: string, configFile: ParsedCommandLine) {
+        const relativePath = getRelativePathFromDirectory(rootDirOfOptions(configFile.options, configFile.options.configFilePath!), inputFileName, /*ignoreCase*/ true);
+        const outputPath = resolvePath(configFile.options.declarationDir || configFile.options.outDir || getDirectoryPath(configFile.options.configFilePath!), relativePath);
+        return changeExtension(outputPath, ".d.ts");
+    }
+
+    function getOutputJavaScriptFileName(inputFileName: string, configFile: ParsedCommandLine) {

```

Comment:
looks like getOwnEmitOutputFilePath. can we consolidate?

Reply:
Holding off on these because `getOwnEmitOutputFilePath` needs a `SourceFile` and an `EmitHost` and we have neither at this point.